fdbserver/StorageCache.actor.cpp

```diff
@@ -1969,15 +1970,73 @@ ACTOR Future<Void> storageCacheServer(StorageServerInterface ssi, uint16_t id, R
 	state StorageCacheData self(ssi.id(), id, db);
 	state ActorCollection actors(false);
 	state Future<Void> dbInfoChange = Void();
+	state StorageCacheUpdater updater(self.lastVersionWithData);
+	self.updater = &updater;
+	state Transaction tr(self.cx);
+	state Value trueValue = storageCacheValue(std::vector<uint16_t>{ 0 });
+	state Value falseValue = storageCacheValue(std::vector<uint16_t>{});
+	state MutationRef privatized;
+	state Version readVersion;
 
 	//TraceEvent("StorageCache_CacheServerInterface", self.thisServerID).detail("UID", ssi.uniqueID).detail("IsCacheServer", ssi.isCacheServer);
+
 	// This helps identify the private mutations meant for this cache server
 	self.ck = cacheKeysPrefixFor( id ).withPrefix(systemKeys.begin);  // FFFF/02cacheKeys/[this server]/
 
 	actors.add(waitFailureServer(ssi.waitFailure.getFuture()));
 	actors.add(traceCounters("CacheMetrics", self.thisServerID, SERVER_KNOBS->STORAGE_LOGGING_DELAY, &self.counters.cc, self.thisServerID.toString() + "/CacheMetrics"));
 
-	// compactCache actor will periodically compact the cache when certain version condition is met
+	TraceEvent(SevDebug, "SCStartUpPostWarmup", self.thisServerID).detail("DataVersionBefore", self.version.get());
+	// Fetch metadata mutation from the database to establish cache ranges and apply them
+	try {
+		loop {
+			tr.setOption(FDBTransactionOptions::READ_LOCK_AWARE);
+			tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			try {
+				Standalone<RangeResultRef> range = wait(tr.getRange(storageCacheKeys, CLIENT_KNOBS->TOO_MANY));
+				ASSERT(!range.more);
+				readVersion = tr.getReadVersion().get();
```

Comment:
I don't see where you pass this read version to the actor that fetches data from the tlog. I think that part is missing and is needed for correctness

Reply:
TLog fetch happens in pullAsyncData(): Line 1729:
cursor = data->logSystem->peekSingle( data->thisServerID, data->version.get() + 1, cacheTag, std::vector<std::pair<Version,Tag>>()) ;

So I set the data->version to be readVersion around line 2026:
self.version.set(readVersion);

But I have found that this produces some test failures. So something is not quite right with passing in readVersion either.

Reply:
I found the problem. This should be fixed in the latest push.